src/unitquaternion.jl

```diff
@@ -0,0 +1,428 @@
+import Base: +, -, *, /, \, exp, log, ≈, ==, inv, conj
+
+abstract type QuatMap end
+
+""" ```
+UnitQuaternion{T,D} <: Rotation
+```
+4-parameter attitute representation that is singularity-free. Quaternions with unit norm
+represent a double-cover of SO(3). The `UnitQuaternion` does NOT strictly enforce the unit
+norm constraint, but certain methods will assume you have a unit quaternion. The
+`UnitQuaternion` type is parameterized by the linearization method, which maps quaternions
+to the 3D plane tangent to the 4D unit sphere. Follows the Hamilton convention for quaternions.
+
+There are currently 4 methods supported:
+* `VectorPart` - uses the vector (or imaginary) part of the quaternion
+* `ExponentialMap` - the most common approach, uses the exponential and logarithmic maps
+* `CayleyMap` - or Rodrigues parameters (aka Gibbs vectors).
+* `MRPMap` - or Modified Rodrigues Parameter, is a sterographic projection of the 4D unit sphere
+onto the plane tangent to either the positive or negative real poles.
```

Comment:
So I've rethought how I was doing this and decided special-casing these to quaternions is misleading and not very general. 

The key idea here is that a rotational error (a difference between two rotations that are reasonably close together) should live in three-dimensional space, the same space as angular velocity.  For globally non-singular representations such as quaternions or rotation matrices we need a map between the higher-dimensional manifold on which they operate to this three-dimensional space. There are several maps that do this. In Lie Group theory the exponential map and Cayley maps are the most common (with the exponential map being the most common, by far). For quaternions, others have proposed using MRPs or the vector part of the quaternion. While the computational expense and global characteristics of these mappings differ (location of singularities) their local behavior are all identical, up to a scaling factor.

In practice, this mapping in incredibly useful for gradient-based optimization methods, where we typically construct local linear or quadratic approximations of nonlinear objectives or constraints, and optimize deviations about some nominal solution iterate. For rotations this means optimizing in the space of 3D rotational errors, which can then be mapped back to a globally nonsingular rotation representation, such as quaternions. This idea has been around in the state estimation community for several decades, in the form of the multiplicative EKF (MEKF or quaternion-multiplicative EKF). 

In practice, we've found that using quaternions together with Rodrigues Parameters (a.k.a. the Cayley Map) works very well, at least for the trajectory optimization and control problems we specialize in. Although RPs are singular at 180 degrees, if your correction step in your optimization algorithm is near 180 degrees, your step is probably too large anyways. They also are nice since they don't have a sign ambiguity like MRPs and are very computationally efficient. 

This might have been more than you asked for but hopefully it clears some stuff up :) I'm going to add some more docstrings that hopefully explain this further.

Reply:
I find state estimation and control fascinating so I love all these details, thanks for explaining!